src/rebuild/initiator.c

```diff
@@ -852,17 +845,76 @@ rebuild_obj_ult(void *data)
 			break;
 	}
 
-	dsc_obj_close(oh);
-free:
 	if (buf != NULL && buf != stack_buf)
 		D_FREE(buf);
+
+	D_DEBUG(DB_REBUILD, "obj "DF_UOID" for shard %u eph "
+		DF_U64"-"DF_U64": rc %d\n", DP_UOID(arg->oid), arg->shard,
+		epr->epr_lo, epr->epr_hi, rc);
+
+	return rc;
+}
+
+static int
+rebuild_obj_punch(struct rebuild_iter_obj_arg *arg)
+{
+	return dss_task_collective(rebuild_obj_punch_one, arg, 0);
+}
+
+/**
+ * Iterate akeys/dkeys of the object
+ */
+static void
+rebuild_obj_ult(void *data)
+{
+	struct rebuild_iter_obj_arg	*arg = data;
+	struct rebuild_pool_tls		*tls;
+	daos_handle_t			oh;
+	daos_epoch_range_t		epr;
+	int				rc;
+
+	tls = rebuild_pool_tls_lookup(arg->rpt->rt_pool_uuid,
+				      arg->rpt->rt_rebuild_ver);
+	D_ASSERT(tls != NULL);
+
+	if (arg->epoch != DAOS_EPOCH_MAX) {
+		rc = rebuild_obj_punch(arg);
+		if (rc)
+			D_GOTO(free, rc);
+	}
+
+	rc = dsc_obj_open(arg->cont_hdl, arg->oid.id_pub, DAOS_OO_RW, &oh);
+	if (rc)
+		D_GOTO(free, rc);
+
+	if (arg->snaps) {
+		int	i;
+
+		for (i = 0; i < arg->snap_cnt; i++) {
+			epr.epr_lo = i > 0 ? arg->snaps[i-1] : 0;
+			epr.epr_hi = arg->snaps[i];
```

Comment:
The EPRs are overlapped, looks it should be [prev_snap + 1, cur_snap], right?

Reply:
ah, right